### ISSUE TITLE
Interruptable occ commands do not get interrupted when pressed ctrl-c

### ISSUE THREAD

**Issue by mmattel** [owncloud/core#22171]
When merging  #22023 a week ago, it seems that a bug silently came in.

Following problem:
When the code for interrupting a occ process was included in scan.php, it worked. After it was taken out and put into the base class (base.php) for reusing for other occ commands, it did not work anymore.

**What works is:**
``if ($this->hasBeenInterrupted()) {``
(https://github.com/owncloud/core/blob/master/apps/files/command/scan.php#L118) calls
``protected function hasBeenInterrupted() {``
(https://github.com/owncloud/core/blob/master/core/command/base.php#L138) triggers
``private function cancelOperation() {``
(https://github.com/owncloud/core/blob/master/core/command/base.php#L153)
here ``$this->interrupted = true`` also works when pressing ctrl-c

**but back in** 
``return $this->interrupted;``
(https://github.com/owncloud/core/blob/master/core/command/base.php#L142)
``$this->interrupted`` is neve true.

I tried it by tracing via command line print the value at each calling level, but when it comes to  function: ``protected function hasBeenInterrupted() {``, ``$this->interrupted`` stays always false independent if ``private function cancelOperation() {`` ``$this->interrupted`` is set to true. 

I did not find the root cause as everything seems to be valid from the coding point of view, but something is missing as it does not work and I do not get it...

BTW, changing ``return $this->interrupted;`` --> ``return true;`` simulates the interrupt correctly.

@DeepDiver1975 @PVince81

**Comment by PVince81** [comment]
@mmattel as usual, PRs are welcome :wink:

**Comment by mmattel** [comment]
@PVince81 sure :smile:

But I need your help to get out of my possible blindness...

As described above, setting ``function cancelOperation()`` ``$this->interrupted = true`` works and the change can be tracked,
but in ``function hasBeenInterrupted()`` ``$this->interrupted`` never changes it state.

For easy handling, I changed the code to call ``hasBeenInterrupted`` only for directories and limit the number of directories to be scanned to 50. (as I am running in my test thru 19k files and >500 directories.) This gives me enough time to press ctrl-c and see the output on the console.

Then I just added the following in both functions (changing the character to be printed accordingly):
````
if ($this->interrupted) {
  print("i");
} else {
  print("n");
}
````
Now I see the change when pressing ctrl-c going on in ``function cancelOperation()`` but not in ``function hasBeenInterrupted()``. This means for me that the interrupt gets properly recognized but somehow the function returning the state has a problem. Why has``$this->interrupted`` not changed its state there (and continues to stay false), it is the same variable??

My request for your support, do you have the same behaviour and if, what can be the cause as the code merged seems to be correct...

**Comment by DeepDiver1975** [comment]
interrupting files:scan using ctrl+c works for me

**Comment by DeepDiver1975** [comment]
please reopen it this still is an issue

**Comment by mmattel** [comment]
Will do.
FYI: yesterday I installed PHPStorm and debugged the two functions line by line.
Monitoring the value ``$this->interrupted``, the situation is exactly as described above. I will try around and look what the cause on my system or in the code could be.

**Comment by mmattel** [comment]
I got some steps further:
My idea was that the different values may come from "different" references in ``$this``. So I did the following:

Checking ``$this`` in ``function configure()`` in ``core/command/base.php``
As you can see below, there are many callers and therefore many references
````
{OC\Core\Command\Status}
{OC\Core\Command\Check}
{OC\Core\Command\App\GetPath}
{OC\Core\Command\App\ListApps}
{OC\Core\Command\App\DeleteConfig}
{OC\Core\Command\App\SetConfig}
{OC\Core\Command\Config\ListConfigs}
{OC\Core\Command\Config\DeleteConfigs}
{OC\Core\Command\Config\System\GetConfigs}
{OC\Core\Command\Config\System\SetConfigs}
{OC\Core\Command\Encryption\ListModules}
{OC\Core\Command\Encryption\Status}
{OC\Core\Command\Security\ListCertificates}
{OC\Core\Command\Scan}
{OCA\Files_External\Command\ListCommand}
{OCA\Files_External\Command\Config}
{OCA\Files_External\Command\Applicapable}
{OCA\Files_External\Command\Import}
{OCA\Files_External\Command\Delete}
````
**Now when you look at:**
![image](https://cloud.githubusercontent.com/assets/3321281/12917739/7b66838a-cf3b-11e5-98b1-68d18d06c651.png)
You see that the reference of ``$this`` is correct ``OC\Core\Command\Scan`` as the function gets called by Scan.

**But when the signal handler gets triggered**
by ``pcntl_signal_dispatch`` you get as reference ``OCA\Files_External\Command\Delete`` which is the last in the list above.
![image](https://cloud.githubusercontent.com/assets/3321281/12917744/813f733e-cf3b-11e5-9fb3-3a7c2422cf7f.png)
And this explains why I can´t interrupt any more. When the logic was in scan.php, ``$this`` was clear.

**Finding summary**
- It seems that function ``cancelOperation`` has the "wrong" reference for ``$this`` 
- It astonishes me that the code does not work for me but works for @DeepDiver1975
- How to fix that ``$that`` will always have the reference of ``hasBeenInterrupted``?
(Or any other method that works for multiple instances)

**Comment by DeepDiver1975** [comment]
For every command (with base class Base) the signal handler is registered.
For current master this will be 18 instances and 18 handlers.

As soon as the interruption occurs all these 18 handlers will be executed.

This explains the 'wrong' reference of $this - as you name it.

What could be a quite valid reason which might explain your observation:
There is a limit of handlers which can be registered.

Idea:

we only setup the signal handler within the execute call

**Comment by DeepDiver1975** [comment]
@mmattel please test https://github.com/owncloud/core/pull/22246

**Comment by mmattel** [comment]
Will do.
Q: do I get it right, that when the ctrl-c event occurs, all handlers are processed, last first (set to true), but because the event gets cleared by this trigger, all other handlers do not get triggered anymore and we get the result as we have seen.

**Comment by nickvergessen** [comment]
The PR will only trigger one handler

**Comment by mmattel** [comment]
Fixed with #22246
